### ISSUE TITLE
Converting API to typescript - steps

### ISSUE THREAD

**Issue by pKorsholm** [medusajs/medusa#724]
We are now converting the API to typescript entirely 

#### Steps to follow in converting endpoints:
- checkout `typescript`
- checkout a new branch `typescript-[store/admin]-[domain]` i.e. `typescript-store-customers` for converting the `api/routes/store/customers` folder
- convert the existing Joi schema to a typescript class:

```
export class AdminCreateUserRequest {
  @IsEmail()
  email: string
  name?: string
  password: string
}
```

- create a response class
```
export type AdminCreateUserResponse = {
  user: Omit<User, "password_hash">
}
```
- use `validator` from `medusa-core-utils` to validate the schema against the request class
```
import { IsEmail } from "class-validator"
import _ from "lodash"
import { validator } from "medusa-core-utils"
import { User } from "../../../.."
import UserService from "../../../../services/user"

export default async (req, res) => {
  const validated = await validator(AdminCreateUserRequest, req.body)

  ...
  res.status(200).json({ user })
}
```
  - _to get specific validation and conversion (such as conversion to numbers) use the [class-validator decorators](https://github.com/typestack/class-validator)_
- update the OAS documentation
  - remember to include `x-authenticated: true` in OAS if the endpoint requires authentication (https://swagger.io/docs/specification/openapi-extensions/)
- update `medusa-js` according to the endpoints 
  - `medusa-js` has been moved from its own repository into `packages/medusa-js` 
- Link `medusa-core-utils` in `packages/medusa`
  - Navigate to `medusa-core-utils` and run `yarn build && yarn link`
  - Navigate to `medusa` and run `yarn link medusa-core-utils`
- Run integration tests
  - Make sure to build both `packages/medusa` and `packages/medusa-core-utils`
  - run `medusa-dev --force-install` in `integration-tests/api`
  - run `yarn test`
- Make pr back into `typescript`

**Comment by srindom** [comment]
Also worth adding some of the agreed-upon naming conventions here:

## Naming API response types
- All API responses should be exported from the "API domain index file" e.g. `api/admin/returns/index.ts`, `api/store/products/index.ts`.
- Responses from the same "API domain" generally follow the same structure so you will likely only have to define a couple of response types for each API domain. E.g. for `api/admin/products/index.ts` you would have: `AdminProductsRes`, `AdminProductsListRes` and `AdminProductsDeleteRes`. The former is the response type for most of the endpoints while the list and delete response are only used for the list and delete endpoints.
- To name the responses you should do `[Admin|Store][subdomain e.g. Products|Returns...][""|List|Delete]Res`

## Naming API request body types
- API request body classes should be exported from the API controller e.g. `api/store/returns/create-return.ts`
- You should name them based on the endpoint's `operationId` (find this in the OAS definition above the API controller) e.g. `PostReturns`. To get the final class name do: `[Admin|Store][operationId]Req` e.g. `StorePostReturnsReq`

## Naming API param types
- API request params classes should be exported from the API controller e.g. `api/store/returns/create-return.ts`
- Use the same approach as for request bodies but this time append `Params` instead of `Req` e.g. `StorePostReturnsParams`

**Comment by olivermrbl** [comment]
Closed #817
